core/templates/dev/head/pages/email_dashboard/EmailDashboardServices.js

```diff
@@ -0,0 +1,120 @@
+ */
+
+oppia.factory('EmailDashboardDataService', [
+  '$http', '$timeout', function($http, $timeout) {
+    var QUERY_DATA_URL = '/emaildashboarddatahandler';
+    var QUERY_STATUS_CHECK_URL = '/querystatuscheck';
+    // No. of query results to display on a single page.
+    var QUERIES_PER_PAGE = 10;
+    // Store latest cursor value for fetching next query page.
+    var latestCursor = null;
+    // Array containing all fetched queries.
+    var queries = [];
+    // Index of currently-shown page of query results.
+    var currentPageIndex = -1;
+
+    var fetchQueriesPage = function(pageSize, cursor) {
+      return $http.get(QUERY_DATA_URL, {
+        params: {
+          num_of_queries_per_page: pageSize,
+          cursor: cursor
+        }
+      }).then(function(response) {
+        return response.data;
+      });
+    };
+
+    return {
+      getQueries: function() {
+        return queries;
+      },
+
+      getCurrentPageIndex: function() {
+        return currentPageIndex;
+      },
+
+      getLatestCursor: function() {
+        return latestCursor;
+      },
+
+      submitQuery: function(data, callback) {
+        var startQueryIndex = (currentPageIndex) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+
+        $http.post(QUERY_DATA_URL, {
+          data: data
+        }).success(function() {
+          // Wait for 500ms and then fetch details of newly submitted query.
+          $timeout(function() {
+            fetchQueriesPage(1, null).then(function(data) {
+              var newQueries = [data.recent_queries[0]];
+              queries = newQueries.concat(queries);
+              callback(queries.slice(startQueryIndex, endQueryIndex));
+            });
+          }, 500);
+        });
+      },
+
+      getNextQueries: function(callback) {
+        var startQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex + 2) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex + 1;
+
+        if (queries.length > endQueryIndex) {
+          callback(queries.slice(startQueryIndex, endQueryIndex));
+        } else {
+          fetchQueriesPage(QUERIES_PER_PAGE, latestCursor).then(function(data) {
+            queries = queries.concat(data.recent_queries);
+            latestCursor = data.cursor;
+            callback(queries.slice(startQueryIndex, endQueryIndex));
+          });
+        }
+      },
+
+      getPreviousQueries: function() {
+        var startQueryIndex = (currentPageIndex - 1) * QUERIES_PER_PAGE;
+        var endQueryIndex = (currentPageIndex) * QUERIES_PER_PAGE;
+        currentPageIndex = currentPageIndex - 1;
+        return queries.slice(startQueryIndex, endQueryIndex);
+      },
+
+      isNextPageAvailable: function() {
+        var nextQueryIndex = (currentPageIndex + 1) * QUERIES_PER_PAGE;
+        return !(queries.length <= nextQueryIndex) || Boolean(latestCursor);
```

Comment:
Should the first clause just be queries.length >= nextQueryIndex? (no need for the negation)